test/videoframeprocessor/DefaultVideoFrameProcessorPipeline.test.ts

```diff
@@ -0,0 +1,432 @@
+
+      class WrongProcessor implements VideoFrameProcessor {
+        name(): string {
+          throw new Error('Method not implemented.');
+        }
+        destroy(): Promise<void> {
+          throw new Error('Method not implemented.');
+        }
+        async process(buffers: VideoFrameBuffer[]): Promise<VideoFrameBuffer[]> {
+          await new Promise(resolve => setTimeout(resolve, (1000 / 15) * 3));
+          return buffers;
+        }
+      }
+      const procs = [new WrongProcessor()];
+      pipe.processors = procs;
+      pipe.setInputMediaStream(mockVideoStream);
+
+      // simulate high latency, has to wait long enough for the callback to be fired.
+      new TimeoutScheduler(800).start(() => {
+        expect(called).to.equal(true);
+
+        pipe.setInputMediaStream(null);
+
+        done();
+      });
+    });
+  });
+
+  describe('getter processors', () => {
+    it('can get processors', () => {
+      const procs = [new NoOpVideoFrameProcessor()];
+      pipe.processors = procs;
+      expect(pipe.processors).to.deep.equal(procs);
+    });
+  });
+
+  describe('stop', () => {
+    it('can stop the processing', async () => {
+      const obs = new MockObserver();
+
+      const procs = [new NoOpVideoFrameProcessor()];
+      pipe.processors = procs;
+      pipe.addObserver(obs);
+      await pipe.setInputMediaStream(mockVideoStream);
+
+      await new Promise(resolve => setTimeout(resolve, 150));
+
+      expect(obs.processingDidStart.called).to.equal(true);
+      pipe.stop();
+      await new Promise(resolve => setTimeout(resolve, 150));
+      expect(obs.processingDidStop.called).to.equal(true);
+    });
+  });
+
+  describe('destroy', () => {
+    it('can stop the processing', async () => {
+      const obs = new MockObserver();
+      const procs = [new NoOpVideoFrameProcessor()];
+      pipe.processors = procs;
+      pipe.addObserver(obs);
+      await pipe.setInputMediaStream(mockVideoStream);
+
+      await new Promise(resolve => setTimeout(resolve, 150));
+
+      expect(obs.processingDidStart.called).to.equal(true);
+      pipe.destroy();
+      await new Promise(resolve => setTimeout(resolve, 150));
+      expect(obs.processingDidStop.called).to.equal(true);
+    });
+
+    it('can destroy processors if they exist', async () => {
+      pipe.processors = null;
+      pipe.destroy();
+
+      class MockProcessor extends NoOpVideoFrameProcessor {
+        destroy = stub();
+      }
+      const obs = new MockObserver();
+      const procs = [new MockProcessor()];
+      pipe.processors = procs;
+      pipe.addObserver(obs);
+      await pipe.setInputMediaStream(mockVideoStream);
+
+      await new Promise(resolve => setTimeout(resolve, 150));
+
+      expect(obs.processingDidStart.called).to.equal(true);
+      pipe.destroy();
+      await new Promise(resolve => setTimeout(resolve, 150));
+      expect(obs.processingDidStop.called).to.equal(true);
+      expect(procs[0].destroy.called).to.equal(true);
```

Comment:
Every time you do this kind of thing you're making our test suite take longer to run.